pkg/lease/lease_test.go

```diff
@@ -0,0 +1,285 @@
+			return errors.Errorf("intervals for replica %s and %s overlap", iv1, iv2)
+		}
+		// Largest possible gap is the 1 extension period + 1 update period. At least
+		// while we don't have randomized transient errors.
+		if gap := iv2.start.Sub(iv1.end); gap > ri.leaseOpts.ExtensionPeriod+ri.leaseOpts.UpdatePeriod {
+			return errors.Errorf("gap of %s between intervals %s and %s too large", gap, iv1, iv2)
+		}
+		iv1 = iv2
+	}
+	return nil
+}
+
+func testLease(t *testing.T, duration time.Duration, leaseOpts *Options, replicaCount int, replicaLifetime time.Duration) []interval {
+	srv := grpc.NewServer()
+	listener := bufconn.Listen(1e6)
+
+	metricServer := &testMetricService{
+		latency:        leaseOpts.RetryPeriod / 10,
+		frequencyLimit: leaseOpts.UpdatePeriod,
+	}
+	monitoring_pb.RegisterMetricServiceServer(srv, metricServer)
+
+	go func() { srv.Serve(listener) }()
+	defer srv.Stop()
+
+	// Run multiple leases simultaniously and wait for a condition violation for the given time.
+	// If there is a leader most of the time and never more than one leader, everything is fine.
+	ctx, cancel := context.WithTimeout(context.Background(), duration)
+	defer cancel()
+
+	bufDialer := func(context.Context, string) (net.Conn, error) {
+		return listener.Dial()
+	}
+	client, err := monitoring.NewMetricClient(ctx,
+		option.WithoutAuthentication(),
+		option.WithGRPCDialOption(grpc.WithInsecure()),
+		option.WithGRPCDialOption(grpc.WithContextDialer(bufDialer)),
+	)
+	if err != nil {
+		t.Fatalf("creating metric client failed: %s", err)
+	}
+
+	var (
+		resource = &monitoredres_pb.MonitoredResource{
+			Type:   "prometheus_target",
+			Labels: map[string]string{"instance": "replica-key"},
+		}
+		intervals = newReplicaIntervals(leaseOpts, time.Now())
+		// Control number of concurrent replicas via channel.
+		replicaCh = make(chan struct{}, 100)
+	)
+	hook := func(name string, start, end time.Time, valid bool) {
+		if valid {
+			intervals.add(name, start, end)
+		}
+	}
+
+Loop:
+	for i := 0; ; i++ {
+		select {
+		case <-ctx.Done():
+			break Loop
+		case replicaCh <- struct{}{}:
+		}
+		// Keep producing new replica IDs. In practice IDs would be recycled but
+		// this makes it easier to debug history.
+		name := fmt.Sprintf("replica-%d", i)
+
+		logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stdout))
+		logger = log.With(logger, "ts", log.DefaultTimestampUTC, "replica", name)
+
+		lease, err := New(nil, "test-project", client, resource, leaseOpts)
+		if err != nil {
+			t.Fatalf("Failed to create lease: %s", err)
+		}
+		lease.Register(func(start, end time.Time, ok bool) {
+			hook(name, start, end, ok)
+		})
+		// Make replicas die after a random amount of time unless a fixed lifetime is given.
+		if replicaLifetime == 0 {
+			replicaLifetime = time.Duration(rand.Intn(int(50 * leaseOpts.ExtensionPeriod)))
+		}
+		t.Logf("starting %s with lifetime %s", name, replicaLifetime)
+
+		ctx, cancel := context.WithTimeout(ctx, replicaLifetime)
```

Comment:
Here you immediately overwrite the timeout context `ctx` initialized above. Is this intentional? As this is in the for loop, it's also creating a chain of contexts with various timeouts. I don't quite follow why...

What's wrong with having a single parent context for all the replicas and timing out?

Reply:
The replicas actually don't live as long as the test. `replicaCount` limits how many there are at a time, but they individually die  after a random amount of time. Otherwise the first replica  to get the lease would just hold it for the entire test duration.

I named it `replicaCtx` so it's not as confusing.